crates/validation_common/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use serde::{Deserialize, Deserializer};
+use std::collections::HashSet;
+
+/// Digest generic in hash size `N`
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Digest<const N: usize>(pub [u8; N]);
+
+impl<'de, const N: usize> Deserialize<'de> for Digest<N> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let dig: String = Deserialize::deserialize(deserializer)?;
+        let dig = hex::decode(dig).map_err(|e| Error::custom(format!("invalid hex: {e}")))?;
+        let dig = dig.try_into().map_err(|v: Vec<_>| {
+            Error::custom(format!(
+                "expected digest to have length of {N}, got {}",
+                v.len()
+            ))
+        })?;
+        Ok(Digest(dig))
+    }
+}
+
+impl<const N: usize> AsRef<[u8; N]> for Digest<N> {
+    fn as_ref(&self) -> &[u8; N] {
+        &self.0
+    }
+}
```

Comment:
Where is the `Borrow` and `Deref` implementations from the diff I sent?
Why not just `git apply`? Are you applying these patches by hand? (you really should not)

Reply:
I didn't know about `git apply`, and I wanted to study what you did, re-type it, in a hope that I'd learn more.

Reply:
Fixed